src/imitation/util/networks.py

```diff
@@ -259,3 +278,99 @@ def build_mlp(
+    name: Optional[str] = None,
+    activation: Type[nn.Module] = nn.ReLU,
+    kernel_size: int = 3,
+    stride: int = 1,
+    padding: Union[int, str] = "same",
+    dropout_prob: float = 0.0,
+    squeeze_output: bool = False,
+    normalize_input_layer: Optional[Type[nn.Module]] = None,
+) -> nn.Module:
+    """Constructs a Torch CNN.
+
+    Args:
+        in_channels: number of channels of individual inputs; input to the CNN will have
+            shape (batch_size, in_size, in_height, in_width).
+        hid_channels: number of channels of hidden layers. If this is an empty iterable,
+            then we build a linear function approximator.
+        out_size: required size of output vector.
+        name: Name to use as a prefix for the layers ID.
+        activation: activation to apply after hidden layers.
+        kernel_size: size of convolutional kernels.
+        stride: stride of convolutional kernels.
+        padding: padding of convolutional kernels.
+        dropout_prob: Dropout probability to use after each hidden layer. If 0,
+            no dropout layers are added to the network.
+        squeeze_output: if out_size=1, then squeeze_input=True ensures that CNN
+            output is of size (B,) instead of (B,1).
+        normalize_input_layer: if specified, module to use to normalize inputs;
+            e.g. `nn.BatchNorm` or `RunningNorm`.
+
+    Returns:
+        nn.Module: a CNN mapping from inputs of size (batch_size, in_size, in_height,
+            in_width) to (batch_size, out_size), unless out_size=1 and
+            squeeze_output=True, in which case the output is of size (batch_size, ).
+
+    Raises:
+        ValueError: if squeeze_output was supplied with out_size!=1.
+    """
+    layers = collections.OrderedDict()
+
+    if name is None:
+        prefix = ""
+    else:
+        prefix = f"{name}_"
+
+    # Normalize input layer
+    if normalize_input_layer:
+        layers[f"{prefix}normalize_input"] = normalize_input_layer(in_channels)
+
+    prev_channels = in_channels
+    for i, n_channels in enumerate(hid_channels):
+        layers[f"{prefix}conv{i}"] = nn.Conv2d(
+            prev_channels,
+            n_channels,
+            kernel_size,
+            stride=stride,
+            padding=padding,
+        )
+        prev_channels = n_channels
+        if activation:
+            layers[f"{prefix}act{i}"] = activation()
+        if dropout_prob > 0.0:
+            layers[f"{prefix}dropout{i}"] = nn.Dropout(dropout_prob)
+
+    # final dense layer
+    layers[f"{prefix}avg_pool"] = nn.AdaptiveAvgPool2d(1)
+    layers[f"{prefix}flatten"] = SmartFlatten()
+    layers[f"{prefix}dense_final"] = nn.Linear(prev_channels, out_size)
+
+    if squeeze_output:
+        if out_size != 1:
+            raise ValueError("squeeze_output is only applicable when out_size=1")
+        layers[f"{prefix}squeeze"] = SqueezeLayer()
+
+    model = nn.Sequential(layers)
+    return model
+
+
+class SmartFlatten(nn.Module):
```

Comment:
Cool module but seems like another piece of complexity that would be eliminated if we just always required a batch dimension.